File: library/kotlin/src/io/envoyproxy/envoymobile/Counter.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import java.lang.ref.WeakReference

/**
 * A stat.
 *
 * Current the supported stat type is counter, and it can increment.
 */
class Counter internal constructor(
  private val envoyEngine: WeakReference<EnvoyEngine>,
  private val elements: List<Element>
) {

  /**
   * Increment the counter by the given count.
   */
  fun increment(count: Int) {
    envoyEngine.get()?.recordCounter(
      elements.joinToString(separator = ".") { it.element }, count
    )
  }
}

File: library/swift/src/stat/Counter.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// A counter, and it can increment
@objc
public class Counter: NSObject {
  private let series: String
  private weak var engine: EnvoyEngine?

  internal init(elements: [Element], engine: EnvoyEngine) {
    self.series = elements.map{ $0.description }.joined(separator: ".")
    self.engine = engine
    super.init()
  }

  func increment(count: Int = 1) {
    guard let engine = self.engine else {
      return
    }

    engine.recordCounter(self.series, count: numericCast(count))
  }
}
